### PR TITLE
Implement RPCs for client_02

### LINKED ISSUE #192: Implement basic RPCs for ICS
- `QueryClientState(id) -> (value, height, proof)` : `height` may be omitted if the response is given in `json`
- `QueryConsensusState(id, height) -> (value, proof)`
- `ComposeHeader(id, height) -> header` : `header` is ok to be bytes. (Relayer doesn't care about the contents)

### PR DESCRIPTION
**Pull request by junha1**
#Closes #192

### CODE REVIEW

**Comment by junha1** [comment]
@majecty Please take a look at the interface. In short, `compose_header()` returns opaque(`Bytes`) data, whereas `query_...()` return transparent(`JSON`) data.

**Comment by majecty** [comment]
@junha1 The interface looks good to me.

File: core/src/ibc/querier/mod.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+/// This module representes all accessible interface
+use crate::ibc;
+use ibc::commitment_23::types::{create_membership_proof, create_non_membership_proof};
+use ibc::commitment_23::{CommitmentPath, CommitmentState};
+use primitives::Bytes;
+use rlp::Decodable;
+
+use crate::ibc::IdentifierSlice;
+
+pub trait Nameable {
```
Comment (majecty) [inline core/src/ibc/querier/mod.rs]: The `Nameable` is used to print a debug message.
How about changing it like `DebugName`?
Then the `name()` should be changed to `debug_name`.

I found a discussion about trait name and it looks good to me.
> If the trait has a single self-explanatory method (or a set of nearly identical methods), name it after the method: Clone, Hash, Default, Into, Write, ToOwned, AsRef, Extend.

https://github.com/rust-lang/api-guidelines/issues/28#issuecomment-418509628

File: rpc/src/v1/types/ibc.rs [hunk 2]
```diff
@@ -13,3 +13,62 @@
...
+/// Because of the nature of IBC, they commonly
+/// 1. Requires a block number for which proof stands
+/// 2. The data should be transparent:
+/// relayer must be able to open it and extract required infomation
+/// 3. Includes a cryptographical proof of that
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct IBCQuery {
+    pub number: u64,
+    pub data: String,
```
Comment (majecty) [inline rpc/src/v1/types/ibc.rs]: Shouldn't we use the data as a generic parameter T?

File: rpc/src/v1/types/ibc.rs [hunk 3]
```diff
@@ -13,3 +13,62 @@
...
+    pub proof: Bytes,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct IBCQueryAbsence {
+    pub number: u64,
+    pub proof: Bytes,
+}
```
Comment (majecty) [inline rpc/src/v1/types/ibc.rs]: When we will use `IBCQueryAbsence` type?
IMO, we should make `IBCQuery.data` as optionable.

**Comment by junha1** [comment]
@majecty I updated all

File: rpc/src/v1/types/ibc.rs [hunk 4]
```diff
@@ -13,3 +13,57 @@
...
+/// Note : proof may represents both that of presence and absence. It depends on option of data.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct IBCQuery<T: Serialize> {
+    pub number: u64,
+    pub data: Option<T>,
+    pub proof: Bytes,
+}
+
+/* -------- Client 02 -------- */
```
Comment (majecty) [inline rpc/src/v1/types/ibc.rs]: We are not using `/*` comment.
Please use `///` and use a full sentence.

File: rpc/src/v1/types/ibc.rs [hunk 5]
```diff
@@ -13,3 +13,57 @@
...
+    /// Unpacked light_client::ClientState
+    pub number: u64,
+    pub next_validator_set_hash: H256,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ConsensusState {
+    pub validator_set_hash: H256,
+    // Unpacked CommitmentRoot
```
Comment (majecty) [inline rpc/src/v1/types/ibc.rs]: Please us `///`

File: rpc/src/v1/types/ibc.rs [hunk 6]
```diff
@@ -13,3 +13,58 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct IBCQuery<T: Serialize> {
+    pub number: u64,
+    pub data: Option<T>,
+    pub proof: Bytes,
+}
+
+/// Client 02 related types
+
```
Comment (junha1) [inline rpc/src/v1/types/ibc.rs]: I added an extra line to say that this comment is not only for the one right after  (`ClientState`).

**Pull request merged**